src/backend/joanie/core/flows/order.py

```diff
@@ -259,41 +200,58 @@ def failed_payment(self):
         Mark order instance as "failed_payment".
         """
 
+    def init(self, billing_address=None):
+        """
+        Transition order to assigned state, creates an invoice if needed and call the flow update.
+        """
+        self.assign()
+        if not self.instance.is_free:
+            if billing_address:
+                Address = apps.get_model("core", "Address")  # pylint: disable=invalid-name
+                address, _ = Address.objects.get_or_create(
+                    **billing_address,
+                    defaults={
+                        "owner": self.instance.owner,
+                        "is_reusable": False,
+                        "title": f"Billing address of order {self.instance.id}",
+                    },
+                )
+
+                # Create the main invoice
+                Invoice = apps.get_model("payment", "Invoice")  # pylint: disable=invalid-name
+                Invoice.objects.get_or_create(
+                    order=self.instance,
+                    defaults={
+                        "total": self.instance.total,
+                        "recipient_address": address,
+                    },
+                )
+            else:
+                raise fsm.TransitionNotAllowed(
+                    "Billing address is required for non-free orders."
+                )
+
+        self.instance.freeze_target_courses()
+        self.update()
+
     def update(self):
         """
         Update the order state.
         """
-        if self._can_be_state_completed():
-            self.complete()
-            return
-
-        if self._can_be_state_to_sign_and_to_save_payment_method():
-            self.to_sign_and_to_save_payment_method()
-            return
-
-        if self._can_be_state_to_save_payment_method():
-            self.to_save_payment_method()
-            return
-
-        if self._can_be_state_to_sign():
-            self.to_sign()
-            return
-
-        if self._can_be_state_pending():
-            self.pending()
-            return
-
-        # TODO: Try to add the following transitions
-        #  if self._can_be_state_pending_payment():
-        #      self.pending_payment()
-        #      return
-        #  if self._can_be_state_no_payment():
-        #      self.no_payment()
-        #      return
-        #  if self._can_be_state_failed_payment():
-        #      self.failed_payment()
-        #      return
-        #  https://github.com/openfun/joanie/pull/801#discussion_r1620640987
+        for transition in [
+            self.complete,
+            self.to_sign,
+            self.to_save_payment_method,
+            self.pending,
+            self.pending_payment,
+            self.no_payment,
+            self.failed_payment,
+        ]:
+            try:
+                transition()
+                return
+            except fsm.TransitionNotAllowed:
+                pass
```

Comment:
Sorry I'm a JS Developer, I like to use more recent things I discover.

```suggestion
            with suppress(fsm.TransitionNotAllowed):
                transition()
                return
```